examples/android/routeguide/app/src/main/java/io/grpc/routeguideexample/RouteGuideActivity.java

```diff
@@ -0,0 +1,402 @@
+                            summary.getFeatureCount(), summary.getDistance(),
+                            summary.getElapsedTime());
+                }
+
+                @Override
+                public void onError(Throwable t) {
+                    failed = t;
+                    finishLatch.countDown();
+                }
+
+                @Override
+                public void onCompleted() {
+                    appendLogs(logs, "Finished RecordRoute");
+                    finishLatch.countDown();
+                }
+            };
+
+            StreamObserver<Point> requestObserver = asyncStub.recordRoute(responseObserver);
+            try {
+                // Send numPoints points randomly selected from the points list.
+                Random rand = new Random();
+                for (int i = 0; i < numPoints; ++i) {
+                    int index = rand.nextInt(points.size());
+                    Point point = points.get(index);
+                    appendLogs(logs, "Visiting point {0}, {1}", RouteGuideUtil.getLatitude(point),
+                            RouteGuideUtil.getLongitude(point));
+                    requestObserver.onNext(point);
+                    // Sleep for a bit before sending the next one.
+                    Thread.sleep(rand.nextInt(1000) + 500);
+                    if (finishLatch.getCount() == 0) {
+                        // RPC completed or errored before we finished sending.
+                        // Sending further requests won't error, but they will just be thrown away.
+                        break;
+                    }
+                }
+            } catch (RuntimeException e) {
+                // Cancel RPC
+                requestObserver.onError(e);
+                throw e;
+            }
+            // Mark the end of requests
+            requestObserver.onCompleted();
+
+            // Receiving happens asynchronously
+            if (!finishLatch.await(1, TimeUnit.MINUTES)) {
+                throw new RuntimeException(
+                       "Could not finish rpc within 1 minute, the server is likely down");
+            }
+            
+            if (failed != null) {
+                throw new RuntimeException(failed);
+            }
+            return logs.toString();
+        }
+    }
+
+    private class RouteChatRunnable implements GrpcRunnable {
+        private Throwable failed;
+
+        @Override
+        public String run(RouteGuideBlockingStub blockingStub, RouteGuideStub asyncStub)
+                throws Exception {
+            return routeChat(asyncStub);
+        }
+        
+        /**
+         * Bi-directional example, which can only be asynchronous. Send some chat messages, and
+         * print any chat messages that are sent from the server.
+         */
+        private String routeChat(RouteGuideStub asyncStub) throws InterruptedException,
+                RuntimeException {
+            final StringBuilder logs = new StringBuilder();
```

Comment:
This is used from multiple threads, so it needs to be thread-safe. `StringBuffer` is thread-safe.